src/main/asciidoc/java/cli.adoc

```diff
@@ -0,0 +1,206 @@
+`java com.acme.Foo`, `com.acme.Foo` is an argument.
+
+Arguments do not have a name, there are identified using a mandatory 0-based index. The first parameter has the
+index `0`:
+
+[source,java]
+----
+CLI cli = CLI.create("some-name")
+    .addArgument(new Argument()
+        .setIndex(0)
+        .setDescription("the first argument")
+        .setArgName("arg1"))
+    .addArgument(new Argument()
+        .setIndex(1)
+        .setDescription("the second argument")
+        .setArgName("arg2"));
+----
+
+The `argName` is optional and used in the usage message.
+
+As options, `link:../../apidocs/io/vertx/core/cli/Argument.html[Argument]` can:
+
+* be hidden using `link:../../apidocs/io/vertx/core/cli/Argument.html#setHidden-boolean-[setHidden]`
+* be mandatory using `link:../../apidocs/io/vertx/core/cli/Argument.html#setRequired-boolean-[setRequired]`
+* have a default value using `link:../../apidocs/io/vertx/core/cli/Argument.html#setDefaultValue-java.lang.String-[setDefaultValue]`
+
+Arguments can also be instantiated from their JSON form.
+
+==== Usage generation
+
+Once your `link:../../apidocs/io/vertx/core/cli/CLI.html[CLI]` instance is configured, you can generate the _usage_ message:
+
+[source,java]
+----
+CLI cli = CLI.create("copy")
+    .setSummary("A command line interface to copy files.")
+    .addOption(new Option()
+        .setLongName("directory")
+        .setShortName("R")
+        .setDescription("enables directory support")
+        .setFlag(true))
+    .addArgument(new Argument()
+        .setIndex(0)
+        .setDescription("The source")
+        .setArgName("source"))
+    .addArgument(new Argument()
+        .setIndex(0)
+        .setDescription("The destination")
+        .setArgName("target"));
+
+StringBuilder builder = new StringBuilder();
+cli.usage(builder);
+----
+
+It generates an usage message like this one:
+
+[source]
+----
+Usage: copy [-R] source target
+
+A command line interface to copy files.
+
+  -R,--directory   enables directory support
+----
+
+If you need to tune the usage message, check the `link:../../apidocs/io/vertx/core/cli/UsageMessageFormatter.html[UsageMessageFormatter]` class.
+
+=== Parsing Stage
+
+Once your `link:../../apidocs/io/vertx/core/cli/CLI.html[CLI]` instance is configured, you can parse the user command line to evaluate
+each option and argument:
+
+[source,java]
+----
+CommandLine commandLine = cli.parse(userCommandLineArguments);
+----
+
+The `link:../../apidocs/io/vertx/core/cli/CLI.html#parse-java.util.List-[parse]` method returns a `link:../../apidocs/io/vertx/core/cli/CommandLine.html[CommandLine]`
+object containing the values. By default, it validates the user command line and checks that each mandatory options
+and arguments have been set as well as the number of values received by each option. You can disable the
+validation by passing `false` as second parameter of `link:../../apidocs/io/vertx/core/cli/CLI.html#parse-java.util.List-boolean-[parse]`.
+
```

Comment:
This is useful if you want to check an argument or option is present even if the parsed command line is invalid.